torch/_dynamo/variables/nn_module.py

```diff
@@ -398,6 +409,66 @@ def generic_call_method_helper(name):
             # Dynamo inlines `__call__`, includes hooks.
             return self.call_function(tx, args, kwargs)
         elif name == "forward":
+            if config.use_single_step_graph:
+                import torch.utils._pytree as pytree
+                from torch._functorch.aot_autograd import create_functional_call
+                from .builtin import BuiltinVariable
+                from .inline_helper import (
+                    decompose_and_inline_function_with_makefx,
+                    reconstruct_node_meta_data,
+                )
+
+                start_node_code = len(tx.output.graph.nodes)
+
+                # For single step graph we want to use make_fx to extract the fx graph
+                # representing the fwd and inline that fx graph.
+                # Dynamo needs to see all inputs and output of each torch function
+                # in case they are saved for the backward.
+                mod = tx.output.get_submodule(self.module_key)
+                fn = mod.forward.__func__
+                # TODO(JackCaoG): considering using source and regualr builder instead of
+                # SourcelessBuilder.
+                fn_source = AttrSource(self.source, "__call__")
+                fn_source = AttrSource(fn_source, "__func__")
+
+                # Create the functional version of the module which also lifted
+                # all parameters as inputs.
+                params = {
+                    **dict(mod.named_parameters(remove_duplicate=False)),
+                    **dict(mod.named_buffers(remove_duplicate=False)),
+                }
+                params_flat, params_spec = pytree.tree_flatten(params)
+                params_flat = list(params_flat)
+                params_len = len(params_flat)
+                functional_call = create_functional_call(
```

Comment:
It's kind of weird that we have to do this, since functional_call IIRC operates by mutating the parameters on the module, which means you'd actually be mutating the live module object during compilation. Very unexpected, and it also feels like it shouldn't be necessary, since we're going to be bytecode evaluating stuff and can just transparently reroute parameter accesses as necessary.

Reply:
TBH I was trying to replicate what aot_autograd did in 
https://github.com/pytorch/pytorch/blob/944d0466451be755f226c379201220d989f45d7e/torch/_functorch/aot_autograd.py#L880-L888

The part I need is to create a new function that takes the lifted parameter on top of the module inputs. Is there a cleaner way to do that?

Reply:
AOTAutograd has to do it this way because it isn't symbolically evaluating bytecode ;) That being said, this is probably fine.....

Reply:
I think it would be nice to avoid the functional_call if that isn't too hard to do.

I imagined that we would want to handle `nn.Linear` in dynamo similar to how user-defined nn modules are handled today (which I am... not extremely familiar with). I imagined that dynamo would burn in the parameters as attributes onto the fx graph, which would then automatically get lifted when AOTAutograd invokes its own `functional_call`

Reply:
hmm ok I can try to remove this code and see how many test will break.